### PR TITLE
Add spec for Floating typeclass

### PR DESCRIPTION
**Pull request by renanroberto**
This PR will help to prevent square root of negative numbers, log(0) or log of negative numbers, raise 0 to a negative number (which would cause division by zero), and add information on the range of some functions, for example sin, cos and tan

### CODE REVIEW

File: tests/neg/cos.hs [hunk 1]
```diff
@@ -0,0 +1,5 @@
+module CosNegTest where
```
Comment (facundominguez) [inline tests/neg/cos.hs]: Thanks @renanroberto for rebasing develop.

Please, use the same name for the modules and the files.

Also, all of the new tests modules need to be added in tests.cabal for CI and our test scripts to pick them.

Lastly, the negative tests need to tell to LH that they are expected to fail with directives like:
```
{-@ LIQUID "--expect-error-containing=AbsNegTest.hs:6:1" @-}
```

See #2158

Comment (renanroberto) [reply tests/neg/cos.hs]: Thanks for the help, @facundominguez! Applied the changes suggested

**Comment by facundominguez** [comment]
Looks like CI encountered some test failures. @renanroberto, the readme contains some instructions to run tests locally, which should fail the same as CI. Let me know if you need help with that.

**Comment by renanroberto** [comment]
@facundominguez I'm having a hard time to identify which test is failing. I can't find it on the CI output and locally the only error I see seems to be unrelated to the PR (see below). Can you show me what is failing so I can try to fix it?

```
[renan@thinkpad liquidhaskell]$ ./scripts/test/test_plugin.sh
Up to date
Running integration tests!
Build profile: -w ghc-9.2.5 -O1
In order, the following will be built (use -v for more details):
 - tests-0.1.0.0 (exe:errors) (first run)
Preprocessing executable 'errors' for tests-0.1.0.0..
Building executable 'errors' for tests-0.1.0.0..
[91 of 91] Compiling ElabLocation3    ( errors/ElabLocation3.hs, /home/renan/haskell/liquidhaskell/dist-newstyle/build/x86_64-linux/ghc-9.2.5/tests-0.1.0.0/x/errors/build/errors/errors-tmp/ElabLocation3.o )

**** LIQUID: ERROR: Sorry, unexpected panic in liquid-fixpoint! ****************

errors/ElabLocation3.hs:187:1: error:
    Uh oh.
    elaborate solver elabBE 928 "lq_anf$##7205759403792805777##d3ap" {lq_tmp$x##8890 : func(0 , [(GHC.Prim.State# GHC.Prim.RealWorld);
                                (Tuple (GHC.Types.TupleRep (GHC.Types.$91$$93$ GHC.Types.RuntimeRep)) GHC.Types.LiftedRep (GHC.Prim.State# GHC.Prim.RealWorld) Tuple)]) | [(lq_tmp$x##8890 =
                                                                                                                                                                              (GHC.Base.$62$$62$ lq_anf$##7205759403792805762##d3aa lq_anf$##7205759403792805776##d3ao))]} failed on:
        lq_tmp$x##8890 == GHC.Base.>> lq_anf$##7205759403792805762##d3aa lq_anf$##7205759403792805776##d3ao
    with error
        Cannot unify (@(711) @(712)) with func(0 , [(GHC.Prim.State# GHC.Prim.RealWorld);
              (Tuple (GHC.Types.TupleRep (GHC.Types.$91$$93$ GHC.Types.RuntimeRep)) GHC.Types.LiftedRep (GHC.Prim.State# GHC.Prim.RealWorld) Tuple)]) in expression: GHC.Base.>> lq_anf$##7205759403792805762##d3aa 
    in environment
        GHC.Base.>> := func(3 , [(@(0) @(1)); (@(0) @(2)); (@(0) @(2))])
    
        lq_anf$##7205759403792805762##d3aa := func(0 , [(GHC.Prim.State# GHC.Prim.RealWorld);
                                                        (Tuple (GHC.Types.TupleRep (GHC.Types.$91$$93$ GHC.Types.RuntimeRep)) GHC.Types.LiftedRep (GHC.Prim.State# GHC.Prim.RealWorld) Tuple)])
    
        lq_anf$##7205759403792805776##d3ao := func(0 , [(GHC.Prim.State# GHC.Prim.RealWorld);
                                                        (Tuple (GHC.Types.TupleRep (GHC.Types.$91$$93$ GHC.Types.RuntimeRep)) GHC.Types.LiftedRep (GHC.Prim.State# GHC.Prim.RealWorld) Tuple)])
    
        lq_tmp$x##8890 := func(0 , [(GHC.Prim.State# GHC.Prim.RealWorld);
                                    (Tuple (GHC.Types.TupleRep (GHC.Types.$91$$93$ GHC.Types.RuntimeRep)) GHC.Types.LiftedRep (GHC.Prim.State# GHC.Prim.RealWorld) Tuple)])
    |
187 | part2 _ = do
    | ^^^^^

```

**Comment by facundominguez** [comment]
> I can't find it on the CI output

If I go to the cabal job of 6ff21a6:
https://app.circleci.com/pipelines/github/ucsd-progsys/liquidhaskell/1790/workflows/65cd8e33-2c0d-49b9-8243-f794845d6948/jobs/8894

And then "Download the full output as a file", and then search there for the string "ERROR:" I can find the same error you are pasting above.

The error itself is complaining that some application of `(>>)` is ill typed. Perhaps reducing the failing test will make the failure easier to understand. 

`@(711) @(712)` stands for a type application of a type variable to another, and it can only unify with some other type application like `Maybe @(1234)`. Instead, a function is being encountered. `func(0 , [A; B])` stands for a function type `A -> B` (which is not represented as `(->) A` applied to `B` for reasons that escape my understanding so far).

**Comment by renanroberto** [comment]
@facundominguez I think I was able to somewhat reduce the problem. On `ElabLocation3.hs` we have the function `part2` where we have this code:

```
let allRounds = iterate {- (Main.round 4 1 m1) -} id example in do
      print $ allRounds !! 1000
      print $ allRounds !! 2000
      print $ allRounds !! 3000
      print $ allRounds !! 4000
      print $ allRounds !! 5000
      print $ allRounds !! 6000
      print $ allRounds !! 7000
      print $ allRounds !! 8000
      print $ allRounds !! 9000
      print $ allRounds !! 10000
```

If instead of 10 `print`s we had 2, the problem would not occur. If we have 3 or more then the problem appear.

I must confess, I have no idea on what is going on. Do you have any advice on how to proceed?

**Comment by facundominguez** [comment]
Thanks @renanroberto. I'll play with this when I have some time.

**Comment by facundominguez** [comment]
I encountered this error while working on something different too. I think it would be fine to remove the `part2` definition from the test.

**Comment by facundominguez** [comment]
@nikivazou, @ranjitjhala,  how do you feel about these specs? Inputs to some functions are allowed to be NaN or Infinity, and I guess that finite precision in the representation of numbers could lead to unexpected results. But perhaps they are still useful to have until something better is proposed.

**Review by nikivazou — APPROVED** [review]

**Comment by nikivazou** [comment]
They look good to me!

**Pull request merged**